src/motion/space_groups.F

```diff
@@ -0,0 +1,1229 @@
+!> \author Pierre-André Cazade (first version)
+! **************************************************************************************************
+   SUBROUTINE spgr_gen(spgr, scoor, types, cell, gopt_env, eps_geo, eps_sym, pol, ranges, &
+                       nparticle, n_atom, n_core, n_shell, iounit)
+      
+      IMPLICIT NONE
+      
+      TYPE(spgr_type)                                    :: spgr
+      TYPE(gopt_f_type), POINTER                         :: gopt_env
+      TYPE(cell_type), POINTER                           :: cell
+      REAL(KIND=dp), DIMENSION(:, :), INTENT(IN)         :: scoor
+      REAL(KIND=dp), DIMENSION(:, :), ALLOCATABLE        :: tmp_coor
+      INTEGER, DIMENSION(:), INTENT(IN)                  :: types
+      INTEGER, DIMENSION(:), ALLOCATABLE                 :: tmp_types
+      INTEGER, DIMENSION(:,:), INTENT(IN), OPTIONAL      :: ranges
+      REAL(KIND=dp), INTENT(IN), OPTIONAL                :: eps_geo, eps_sym
+      REAL(KIND=dp), DIMENSION(3), OPTIONAL              :: pol
+      INTEGER, OPTIONAL                                  :: nparticle, n_atom, n_core, n_shell
+      INTEGER, INTENT(IN), OPTIONAL                      :: iounit
+
+      CHARACTER(LEN=*), PARAMETER :: routineN = 'spgr_gen', routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: handle, ierr, major, micro, minor, &
+                                                            nop, tra_mat(3, 3), i, j, n_sr_rep
+      LOGICAL                                            :: spglib
+
+      CALL timeset(routineN, handle)
+
+      !..total number of particles (atoms plus shells)
+      IF (PRESENT(nparticle)) THEN
+         CPASSERT(nparticle==SIZE(scoor, 2))
+         spgr%nparticle = nparticle
+      ELSE
+         spgr%nparticle = SIZE(scoor, 2)
+      ENDIF
+
+      IF (PRESENT(n_atom)) THEN
+         spgr%n_atom = n_atom
+      ELSE IF (PRESENT(n_core)) THEN
+         spgr%n_atom = spgr%nparticle - n_core
+      ELSE IF (PRESENT(n_shell)) THEN
+         spgr%n_atom = spgr%nparticle - n_shell
+      ELSE
+         spgr%n_atom = spgr%nparticle
+      ENDIF
+
+      IF (PRESENT(n_core)) THEN
+         spgr%n_core = n_core
+      ELSE IF (PRESENT(n_shell)) THEN
+         spgr%n_core = n_shell
+      ELSE
+         spgr%n_core = 0
+      ENDIF
+
+      IF (PRESENT(n_shell)) THEN
+         spgr%n_shell = n_shell
+      ELSE IF (PRESENT(n_core)) THEN
+         spgr%n_shell = n_core
+      ELSE
+         spgr%n_shell = 0
+      ENDIF
+
+      IF ( .NOT.( spgr%nparticle == (spgr%n_atom + spgr%n_shell) ) ) THEN
+         CPABORT("spgr_gen: nparticle not equal to natom + nshell.")
+      END IF
+
+      spgr%nparticle_sym = spgr%nparticle
+      spgr%n_atom_sym = spgr%n_atom
+      spgr%n_core_sym = spgr%n_core
+      spgr%n_shell_sym = spgr%n_shell
+
+      ! output unit
+      IF (PRESENT(iounit)) THEN
+         spgr%punit = iounit
+      ELSE
+         spgr%punit = 6
```

Comment:
hardcoded print units are difficult to maintain, you can use `get_unit_number` or `open_file` from the `cp_files` module instead to get dynamic consistent unit numbers (or files)